### PR TITLE
ID3v2 key notation export 

### LINKED ISSUE #10890: Wrong Key format written to the file Tags
### Bug Description

According to https://id3.org/id3v2.3.0 the content of TKEY is well defined: 

>TKEY
>    The 'Initial key' frame contains the musical key in which the sound starts. It is represented as a string with a maximum length of three characters. The ground keys are represented with "A","B","C","D","E", "F" and "G" and halfkeys represented with "b" and "#". Minor is represented as "m". Example "Cbm". Off key is represented with an "o" only. 

Mixxx currently writes the String visible at the GUI into the TKEY field. This is wrong and can cause issues in third party applications. 

### Version

2.3.3 and main 

### OS

All

### PR DESCRIPTION
**Pull request by daschuer**
This makes sure that newly edited keys are written with the ID3v2 format to the database and track metadata. 
Before the current selected key notation from the preferences is used. 
This is likely incompatible to other tools. 

Fixes #10890

This PR is on top of https://github.com/mixxxdj/mixxx/pull/11096

### CODE REVIEW

**Comment by daschuer** [comment]
My original idea was to make this a part of 2.3.4. 
But it has grown a bit because of the required refactoring as a prerequisite.  
Now I am unsure. This is an important fix, to write the correct format and it has a unit-test.
Ideas?

File: src/library/dao/trackdao.cpp [hunk 1]
```diff
@@ -1313,9 +1308,7 @@ TrackPointer TrackDAO::getTrackById(TrackId trackId) const {
         TrackPopulatorFn populator = columns[i].populator;
         if (populator != nullptr) {
             // If any populator says the track should be dirty then we dirty it.
-            if ((*populator)(queryRecord, i, pTrack)) {
-                shouldDirty = true;
-            }
+            shouldDirty = (*populator)(queryRecord, i, pTrack);
```
Comment (uklotzde) [inline src/library/dao/trackdao.cpp:1311]: Why did you change this unrelated code and introduce a bug? It no longer works as intended.

Comment (daschuer) [reply src/library/dao/trackdao.cpp:1311]: Ah, OK, I see. I will revert this. Thank you.

Comment (uklotzde) [reply src/library/dao/trackdao.cpp:1311]: Please note that I discovered this flaw by chance and didn't do a review of the remaining code.

**Comment by atskler** [comment]
After this fix what file formats and what kind of tags will be affected at the end?
For example an mp3 with ape tag will get the new notation format? And mp4 tags? flac tags?
For general consistency and uniformity should I rewrite tags for all my files?
Or just specific formats with specific tags?

Will be mixed notation info in Mixxx's database or that will be updated to be uniform and consistent?

**Comment by daschuer** [comment]
After merging this PR, Mixxx reads the key field from all tag types literally to the database. The GUI shows the key as you specify it in the key preferences. When you edit a key value or reanalyse the track, the key is stored with the ID3v2 notation for all tag types.   
This is a most conservative approach that keeps the file metadata unchanged along with possible extra infos like tuning or special modes. It does not fix the non-standard key notation written by the current Mixxx.  

Does that work for you?

**Comment by atskler** [comment]
If i'm right all this started with "can cause issues in third party applications" (10890). And if we are talking about issues with third party applications, then I would like to have the option to "fix the non-standard key notation written by the current Mixxx" for them, additionally because I would not like to see mixed notation sourced from tags in a 3rd party application like foobar2000.

**Comment by daschuer** [comment]
Yes, right. What is your preferred solution?

**Comment by atskler** [comment]
Provide an option in preferences to choose:
- legacy notation scheme (idiosyncratic, less compatible)
- standardized notation scheme (fully compatible)

Also provide a function to rewrite all tags accordingly.

On new installations use the standardized scheme by default.
On existing installations use the legacy scheme by default.

What do you think?

**Comment by daschuer** [comment]
I consider writing the legacy format a bug. So I don't think it is reasonable to provide an option for this format. Do you have a use case for this? 

We have currently no good way to mass update all Tracks. This will take loooong and we need to implement a reasonable feedback for this. Would it help to update only touched tracks? This way you could be forced by analysis or manual via the "Export Metadata to file tags".

There is by the way the risk to loose the tuning information written by Rapid Evolution.

**Comment by atskler** [comment]
Then consider that on my side Mixxx implanted bug into 6072 files, and all those files needs to be fixed if we really have the fear regarding those lurking "issues in third party applications".

"Do you have a use case for this?" - Yes, a general player can't display this or that notation format, just show what is written into the tag, and for example for me the 12B, 9A format is more convenient. If you fix this, you actually brake my "way of life" and there is no way back.

Regarding the Rapid Evolution risk you mentioned: it sounds like it also uses some non-standard idiosyncratic "bug" format. Otherwise why could be any problem here? And if there is implemented compatibility for a Rapid Evolution "bug", then a backwards compatibility for the other "buggy" notation format may be also implemented. (Also I found that Rapid Evolution is discontinued, abandoned, unsupported software.)

At the and, in my opinion, you have to provide fix for the files also, not just for the code, if you really want to mend this bug, and that fear.

How do you see all this?

**Comment by atskler** [comment]
Anyways as I see it currently Mixxx not reads/imports back the key info from the mp3 tags, at lest in the format of “12A”. Probably the proposed standard way should be implemented, and then the tags rewritten. But then the standard should be precisely followed regardless of the compatibility with same abandoned software that again does non-standard things, or a special compatibility mode should be provided for them in preferences, disabled by default.

**Comment by daschuer** [comment]
Sounds good. I think that is nothing we should do in a stable branch. I will re-target it to 2.4.

**Comment by fwcd** [comment]
Could you clarify which key notation you are writing here? Haven't dug into the code, but IIRC ID3v2 doesn't define a standard key notation (or even tag), so I'll assume that traditional notation (e.g. "D♭") is meant?

**Comment by daschuer** [comment]
This one is obsolete because an alternative has already been merged.
